add Attendance model, controller, api, swagger; DateValidation invalid_date_msg

weedySeaDragon opened the pull request:
# 💅🏼 What issue does this fix?
<!-- Which Github Issue is this related to?  Summarize the work in a sentence or two -->
#163 

## 🍂 Before You Submit
<!-- Check steps as necessary - this list is a reminder -->
* [x] Did you write tests?
* [x] Did you run `bundle exec rspec` from the root?
* [x] Did you run `bundle exec rails rswag` from the root?
* [x] Did you run `bundle exec rubocop` from the root?
* [x] Did you run `yarn lint` in `/client`?
* [x] Did you run `yarn test` in `/client`?
* [x] Are your primary keys UUIDs on any new tables?

## 🛷 Deployment Considerations
<!-- What do we need to know to deploy this code out? -->
* [ ] Data Migrations
* [x] Schema Migrations
* [ ] Dependencies

## 🧵 Steps to set up locally

<!--
A list of things you need to change to get the code going
* Any new environment variables
* Any build steps
* Any docker changes
* Any migrations or tasks that need to run manually
-->

## 🧳 Steps to test
<!-- Outline how to confirm the changes. Very similar to the **Steps to Reproduce** from tickets -->

## 🕯 Caveats, concerns
<!-- Anything you'd like to bring to the attention of reviewers -->

I added a public method to DateParamValidator to access the _invalid date message_ so it can be used in RSpecs.

katelovescode commented on app/models/attendance.rb:
```diff
@@ -0,0 +1,20 @@
+class Attendance < UuidApplicationRecord
+  belongs_to :child_case_cycle
+
+  LENGTHS_OF_CARE = %w[part_day full_day full_plus_part_day full_plus_full_day].freeze
```
@csprayregen when someone adds an attendance, are we expecting them to choose what length of care it is, or are we calculating that by the number of hours as laid out in their subsidy rule? 

If we're calculating it, can the user change it, if for some reason our calculation gets it wrong?

csprayregen replied:
Since we haven't designed these screens yet, I don't think we have clear answers. My assumption, though, is that we're calculating it for them based on the number of hours, as laid out in their subsidy rule. I'm not sure if we'll let them change it.

weedySeaDragon replied:
I'll put a note in the model so these comments are captured.

katelovescode reviewed (commented):
Outstanding business logic question for Chelsea

katelovescode commented on app/models/attendance.rb:
```diff
@@ -0,0 +1,57 @@
+#   If we're calculating it, can the user change it, if for some reason our
+#   calculation gets it wrong?
+#
+#   2020-09-08: Answer from Casey S:
```
```suggestion
#   2020-09-08: Answer from Chelsea S:
```

weedySeaDragon replied:
whoops

katelovescode commented on app/controllers/api/v1/attendances_controller.rb:
```diff
@@ -0,0 +1,56 @@
+  # Only allow a trusted parameter "white list" through.
+  def attendance_params
+    params.require(:attendance).permit(
+      :id, :child_case_cycle_id, :slug, :length_of_care, :starts_on
```
Per the conversation with Chelsea, can we remove length of care from the allowed params? And maybe put a TODO  in new/create to calculate length of care based on subsidy rules?

weedySeaDragon replied:
will do.

katelovescode commented on db/migrate/20200907004651_create_attendances.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :attendances, id: :uuid do |t|
+      t.references :child_case_cycle, type: :uuid, null: false, foreign_key: true
+
```
We missed two fields in the data model; `check_in_time` and `check_out_time` - waiting for confirmation from Chelsea if that's the right name for the fields, but they should both be `TIME` type fields, formatted only to the minute.  We have some open questions so I think we should keep this PR on hold until we get them answered.

Open questions:

- Is check_in_time and check_out_time the right way to do it?
- Also are those times rounded for calculation of length of care?  Or are they to-the-minute?

katelovescode replied:
This model should have `starts_on`, `check_in`, `check_out`, `total_time_in_care`, `slug`, and let's rename `length_of_care` to `attendance_duration`.  We'll do the rounding/logic in a later ticket.

katelovescode replied:
Interesting:  https://www.postgresqltutorial.com/postgresql-interval/  I've never used this data type before.  This might suit our needs for `total_time_in_care`?

weedySeaDragon replied:
Cool.  postgres `interval` + Ruby `Duration` = yes, that looks like it'd work!

weedySeaDragon commented:
@katelovescode :  I made the changes to Attendance as you asked.
I also added in a relation to `child_site` because that seems like it _really_ needs it. 

Annnd  I added in a bunch of code into `db/seeds.rb` so that Attendance records are created and seeded.   There are lots of ways to organize code used to seed things (make some 'helper files' to just keep the code better organized, etc etc.  I left this as one big honkin' file for now; we (you) can determine how you want to deal with it.  

I wanted to just go ahead and get this in so that we can then start noodling/percolating on modeling the _lengths_of_care_ and _subsidy_rules_ and _definitions of things_  etc.

katelovescode commented on app/models/attendance.rb:
```diff
@@ -17,41 +17,58 @@
+  has_one :child, through: :child_site
+  has_one :site, through: :child_site
+
+  belongs_to :child_case_cycle # TODO: really?
```
I think so; we could do the calculation based on dates but I think getting the indexing to its child_case_cycle for free might help with performance later.

katelovescode commented on app/validators/time_param_validator.rb:
```diff
@@ -0,0 +1,16 @@
+  def validate_each(record, attribute, value)
+    value&.is_a?(Time) ? value : Time.zone.parse(value)
+  rescue TypeError, ArgumentError
+    record.errors.add(attribute, self.class.invalid_time_msg)
```
❤️

katelovescode commented on app/models/attendance.rb:
```diff
@@ -17,41 +17,58 @@
+
+  belongs_to :child_case_cycle # TODO: really?
 
   LENGTHS_OF_CARE = %w[part_day full_day full_plus_part_day full_plus_full_day].freeze
```
Rename to `DURATION_DEFINITIONS`?

weedySeaDragon replied:
will do

katelovescode commented on db/migrate/20200913004651_create_attendances.rb:
```diff
@@ -1,10 +1,14 @@
       t.date :starts_on, null: false
+      t.time :check_in, null: false
+      t.time :check_out, null: false
+      t.interval :total_time_in_care, null: false, comment: 'Calculated: check_out time - check_in time'
```
😱 😍

katelovescode commented on app/models/attendance.rb:
```diff
@@ -17,41 +17,58 @@
   end
+
+  def calc_total_time_in_care
+    self.total_time_in_care = (check_out.nil? || check_in.nil? ? 0 : check_out - check_in)
```
👍

katelovescode commented on db/seeds.rb:
```diff
@@ -22,13 +37,46 @@
+
+# find_or_create_by! a Child with the full_name,
+#  and birthday set randomly between the min_age and max_age.
+def child_named(full_name, min_birthday: (Time.zone.now - 2.weeks),
```
Can we extract these min_birthday and max_birthday into constants above?  These might come in handy later.

katelovescode commented on spec/models/child_site_spec.rb:
```diff
@@ -5,6 +5,8 @@
   it { should belong_to(:child) }
   it { should belong_to(:site) }
+
+  # TODO: validate started_care, ended_care
```
I don't think this will be required for now, so the only validations I can think of are date validations

katelovescode reviewed (commented):
Looks great, just a couple comments/requests
